sheepdog/transactions/upload/sub_entities.py

```diff
@@ -0,0 +1,454 @@
+
+        # ################################################################
+        # SignpostClient is used instead of IndexClient for the GDCAPI.
+        # This means that the client doesn't have access to IndexClient's
+        # methods, causing exceptions to occur.
+        #
+        # Temporary workaround until gdcapi uses indexd
+        # ################################################################
+        if self._config.get('USE_SIGNPOST', False):
+            file_by_uuid = self.get_file_from_index_by_uuid(uuid)
+            if not file_by_uuid:
+                self.record_error(
+                    "Could not find file in index with ID: {}".format(uuid),
+                    type=EntityErrors.INVALID_VALUE
+                )
+                is_valid = False
+        else:
+            file_by_hash = self.get_file_from_index_by_hash()
+            file_by_uuid = self.get_file_from_index_by_uuid(uuid)
+
+            if not file_by_hash or not file_by_uuid:
+                error_message = (
+                    'Could not find exact file match in index for id: {} '
+                    'AND `hashes - size`: `{} - {}`. '
+                ).format(
+                    self.entity_id,
+                    str(self._get_file_hashes()),
+                    str(self._get_file_size())
+                )
+
+                if file_by_hash:
+                    error_message += 'A file was found matching `hash / size` but NOT id.'
+                elif file_by_uuid:
+                    error_message += 'A file was found matching id but NOT `hash / size`.'
+                else:
+                    # keep generic error message since both didn't result in a match
+                    pass
+
+                self.record_error(
+                    error_message,
+                    type=EntityErrors.INVALID_VALUE
+                )
+                is_valid = False
+
+            if file_by_hash and file_by_uuid and file_by_hash.did != file_by_uuid.did:
+                # both exist but dids are different
+                # FIXME: error should be handled different/removed
+                #        when we support updating indexed files
+                self.record_error(
+                    'Provided id for indexed file {} does not match the id '
+                    'for the file discovered in the index by hash/size ('
+                    'id: {}). Updating a previous index with new file '
+                    'is currently NOT SUPPORTED.'
+                    .format(file_by_uuid.did, file_by_hash.did),
+                    type=EntityErrors.INVALID_VALUE,
+                )
+                is_valid = False
+
+        return is_valid
+
+    def get_file_from_index_by_hash(self):
+        """
+        Return the record entity from "signpost" (index client)
+
+        NOTE:
+        - Should only ever be called for data and metadata files.
+        - If there is already a record matching the hash and size for this
```

Comment:
what does this mean?

Reply:
removed in 69d1e4f. was a copy and paste artifact from before. just clarifies the usage but isn't really needed in the subclass now